reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+
+```html
+<template id="t-combobox-input">
+  <input id="real-input" role="combobox" type="text" />
+</template>
+
+<script>
+customElements.define("combobox-input", 
+  class ComboboxInput extends HTMLElement {
+    static observedAttributes = ["listbox"];
+
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-combobox-input").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-input"; // (1)
+      this.input_ = this.shadowRoot_.getElementById('real-input');
+    }
+
+    attributeChangedCallback(attr, _oldValue, value) {
+      if (attr === "listbox") {
+        // (2)
+        const listbox = this.getRootNode().getElementById(value);
+        this.input_.ariaControlsElements = [listbox];
+        this.input_.ariaActiveDescendantElement = listbox;
+      }
+    }
+  });
+</script>
+```
+
+#### `<combobox-listbox>`
+
+This component is a wrapper around `<div role="listbox">` and the `<div role="option">` items inside.
+
+1. It delegates all references _except_ `aria-activedescendant` to the `<div role="listbox">` using `referenceDelegate`.
+2. It has a custom attribute `activeitem`, which is used to control which item gets the `aria-activedescendant` delegation using `referenceDelegateMap`. This lets the parent component control the active item.
+
+```html
+<template id="t-combobox-listbox">
+  <div id="real-listbox" role="listbox">
+    <div id="option-1" role="option">Option 1</div>
+    <div id="option-2" role="option">Option 2</div>
+  </div>
+</template>
+
+<script>
+customElements.define("combobox-listbox", 
+  class ComboboxListbox extends HTMLElement {
+    static observedAttributes = ["activeitem"];
+
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-combobox-listbox").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-listbox"; // (1)
+    }
+
+    attributeChangedCallback(attr, _oldValue, value) {
+      if (attr === "activeitem") {
+        this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = value; // (2)
+      }
+    }
+  });
+</script>
+```
+
+#### `<fancy-combobox>`
+
+This component combines the two subcomponents above into a combobox.
+
+1. It hooks up the listbox to the input using the `<combobox-input>`'s custom `listbox` attribute.
+2. It controls which item inside the listbox is the `aria-activedescendant` of the input using the `<combobox-listbox>`'s custom `activeitem` attribute.
+3. It delegates all references to the `"combo-input"` component, which itself delegates to the `"real-input"` inside.
+
+```html
+<template id="t-fancy-combobox">
+  <combobox-input id="combo-input" 
+                  listbox="combo-listbox"> <!-- (1) -->
```

Comment:
I wish there was a quality path towards non-normative API here, so that `<fancy-input>` could be consumed as a drop-in replacement of `<input>` in something like [the ARIA APG demos for combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-both/#rps_label_textbox).